doc/tools/docgram/README.md

```diff
@@ -0,0 +1,194 @@
+Grammar extraction tool for documentation
+=========================================
+
+doc_grammar extracts Coq's grammar from .mlg files, edits it and inserts it in
+chunks into .rst files.  The current prototype generates Sphinx
+`productionlist` displays.  For `tacn` and `cmd`, the tool verifies that the
+.rst matches the grammar.  Updates to tactics and commands must be made
+manually since the grammar doesn't have names for them as it does for
+nonterminals.  `tacv` and `cmdv` are not processed at all.
+
+The mlg grammars present several challenges to generating an accurate grammar
+for documentation purposes:
+* The 30+ mlg files don't define an overall order in which nonterminals should
+appear in a complete grammar.
+* Even within a single mlg file, grammar rules and productions are often given
+in an order that's much different from what a reader of the documentation would
+expect.  In a small number of cases, the order used is necessary for technical
+reasons.
+* For a few nonterminals, some productions are only available if certain plugins
+are activated (e.g. SSR).  Readers should be informed about these.
+* Some limited parts of the grammar are defined in OCaml, including lookahead symbols
+like `test_bracket_ident` and references to nonterminals in other files using
+qualified names such as `Prim.ident`.  A few symbols are defined multiple times,
+such as `scope` and `orient`.
+
+What the tool does
+------------------
+1. The tool reads all the mlg files and generates `fullGrammar`, which includes
+all the grammar without the actions for each production or the OCaml code.  This
+file is provided as a convenience to make it easier to examine the (mostly)
+unprocessed grammar of the mlg files with less clutter.  Nonterminals that use
+levels ("5" RIGHTA below) are modified, for example:
+
+```
+  tactic_expr:
+    [ "5" RIGHTA
+      [ te = binder_tactic -> { te } ]
+```
+
+becomes
+
+```
+tactic_expr5: [
+| binder_tactic
+| tactic_expr4
+]
+```
+
+2. The tool applies grammar editing operations specified by `common.edit_mlg` to
+generate `editedGrammar`.
+
+3. The tool uses `orderedGrammar` to give the desired order for the nonterminals
+and productions in the documented grammar.  `orderedGrammar` has
+the same format as `editedGrammar`.  The tool verifies that `orderedGrammar` is
+a permutation of `editedGrammar`.  Developers are responsible for editing
+`orderedGrammar`.
```

Comment:
Sorry to insist. I keep looking for a solution that will avoid editing such a file manually for most grammar updates. Here is a proposal: split the reordering in two parts.

- First part: reorders only the non-terminals. For this part, we only need the list of non-terminal in a certain order (and it is not necessary to edit when changing the grammar without changing the list of non-terminals).
- Second part: reorders only the production rules for the non-terminals which need their production rules reordered. The win here is that the non-terminals that do not need their rules reordered are simply unlisted in this second part.

Reply:
I agree that having to update orderedGrammar by hand is tedious.  It's just the least-bad solution I could come up with.  I'm certainly open to better ideas.

What you propose here would indeed reduce the amount of data required, but I think it would be more tedious to edit.  If I'm adjusting the order of the nonterminals, that generally depends on seeing all the productions, e.g. if I see `foo : bar + baz` I would think about putting the NTs in the order `foo, bar, baz`.  But I won't see the right hand sides in the NTs file you suggest.  I would have to look at, say, `editedGrammar` for the RHSs, then copy only the NT order to the other file.  If I then want to see the resulting grammar, I have to regenerate `editedGrammar`.  And ordering the grammar is frequently an iterative process.

The same thing applies to reordering the production rules.  Also, the NTs in the second file may not be in the order they appear in the NTs file.  That adds a bit more complexity for the user.

In contrast, the current orderedGrammar is WYSIWYG.

Also, I don't expect anyone besides me any maybe you will touch orderedGrammar until we get the grammar fully into the .rst files.  I think it's something we could change later.  It's not clear there's a big win form changing it right now.

What do you think?

Reply:
I agree that this is something we can refine later, and keep this way in the current PR.